### PR TITLE
feat: Add the stop condition to allow users to stop based on offsets

### PR DESCRIPTION
**Pull request by palmere-google**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-pubsublite-flink/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

### CODE REVIEW

File: src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSourceSettings.java [hunk 1]
```diff
@@ -94,6 +95,38 @@
 
   abstract @Nullable SerializableSupplier<CursorClient> cursorClientSupplier();
 
+  abstract @Nullable SerializableSupplier<TopicStatsClient> topicStatsClientSupplier();
+
+  AdminClient getAdminClient() {
+    if (adminClientSupplier() != null) {
+      return adminClientSupplier().get();
+    }
+    return AdminClient.create(
```
Comment (jiangmichaellll) [inline src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSourceSettings.java:104]: I think you could provide this as the default value in constructor, then you don't need nullable and you could avoid the null check. Same applies to below methods.

File: src/main/java/com/google/cloud/pubsublite/flink/internal/reader/MaybeCompleteSubscriberFactory.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.cloud.pubsublite.SequencedMessage;
+import com.google.cloud.pubsublite.flink.internal.reader.PartitionFinishedCondition.Result;
+import com.google.cloud.pubsublite.flink.internal.split.SubscriptionPartitionSplit;
+import com.google.cloud.pubsublite.internal.CheckedApiException;
+import java.util.Optional;
+
+public class MaybeCompleteSubscriberFactory implements CompletablePullSubscriber.Factory {
```
Comment (jiangmichaellll) [inline src/main/java/com/google/cloud/pubsublite/flink/internal/reader/MaybeCompleteSubscriberFactory.java:26]: ConditionalCompletableSubscriberFactory?
Also some comments would be great.

File: src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+public class StopCondition implements Serializable {
+  enum Condition {
+    INDEFINITE,
```
Comment (jiangmichaellll) [inline src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java:39]: nit some comments for each Condition would be great.

File: src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    HEAD,
+    OFFSETS,
+  }
+
+  private final Condition condition;
+  private final Map<Partition, Offset> finalSplits;
```
Comment (jiangmichaellll) [inline src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java:45]: I would suggest having 3 subclasses of StopCondition for each condition case, and each handles those tofinishedCondition separately, also you would embed getHeadOffsets only in HEADStopCondition. I think that would be clearer.

**Review by jiangmichaellll — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java (hunk 3 above)
Comment (palmere-google) [reply src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java:39]: Conditions are now gone

File: src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java (hunk 4 above)
Comment (palmere-google) [reply src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java:45]: I like that, done

File: src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSourceSettings.java (hunk 1 above)
Comment (palmere-google) [reply src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSourceSettings.java:104]: Unfortunately this doesn't work :(

This method depends on the subscription path, which isn't available when we set the default values.

Comment (jiangmichaellll) [reply src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSourceSettings.java:104]: I see, SG.

File: src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+  private static class ReadToOffsets extends StopCondition {
+    private final Map<Partition, Offset> offsets;
+
+    private ReadToOffsets(Map<Partition, Offset> offsets) {
+      this.offsets = offsets;
+    }
+
+    @Override
+    public StopCondition canonicalize(
```
Comment (jiangmichaellll) [inline src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java:70]: I think you could provide a default implementation to return this, and only override ReadToHead.

**Review by jiangmichaellll — APPROVED** [review]
LGTM.

File: src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java (hunk 5 above)
Comment (palmere-google) [reply src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java]: Done

**Pull request merged**